### ISSUE TITLE
[Feature request] Modulation

### ISSUE THREAD

**Issue by Kawaxte** [midilab/jc303#19]
Hi!

I'm coming from the likes of Image-Line TS404 Bassline Synthesizer, unfortunately deprecated piece of amazing but cheap TB-303 emulation with built-in modulation and great distortion.

Perhaps there could be a setting, like on Arturia Acid V, to have modulation settings bindable to things like Cutoff?

**Comment by midilab** [comment]
Hey @Kawaxte,

I had some great times with TS404 during the FL4 era.

This project is a port of the Open303 engine (which uses the Open303 as the main DSP library) to make it available under different plugin formats and OS platforms using Juce. So, anything other than that is out of this project's scope for now, at least until we complete all the Roadmap features. As much as I would like to see some kind of TS404 open-sourced, this project won't touch the Open303 source code.

You mentioned TS404, that has an LFO for modulation, and Acid V, that has a Vibrato (which is technically an LFO as well). Maybe I can explore some kind of modulation setup in the future.

Thanks for the suggestion. It's marked as an enhancement issue, and it could go to development after all the Roadmap items are completed (see README.md).

On the other hand, I am currently working on a redesign and overdrive/distortion features. 

Since the Open303 source code is used as a library, with no changes planned on his src code, I decided to try integrating another open-source plugin, called NeuralPi(which uses AI to model real distortion guitar pedals), into chain mode with Open303's output sound going to the distortion plugin with dry/wet setup. So far, the results are promising, using TS9 and Little Big Muff models, but the idea is to let user chose the model for JC303 plugin, by importing new ones or selecting factory ones.
https://guitarml.com/tonelibrary/tonelib-npi.html

**Comment by RobinSchmidt** [comment]
Cool! Looking forward to see the amp sim integrated. Distortion is the essential effect on a 303 sound. By the way - I recently stumbled over this:

https://www.acidvoice.com/tb_303_best_distortion_pedal.htm

which is very cool

**Comment by midilab** [comment]
@RobinSchmidt , great to have you around!

Indeed 303 and a overdrive is a must!

I know this website! i use it as main source of a research for the ideal overdrive to integrate with jc303. 

The idea is to use guitarml model files for import presets, they have some great models but i decide to release first version with TS9 overdrive model before implement the import and load models directly on the plugin.

The TS9 gutarml model does a great job! i have tested most of opensource guitar pedals plugins, from algorithm to modeling ones, and at the end the best implementation mostly because of optmizations for plataform code generation and the sound quality is the [BYOD](https://github.com/Chowdhury-DSP/) guitarml implementation.

Other good thing to keep a loadable schema of models is that the community could generate new models maybe based on the best overdrives of acidvoice website for example. Or anyone that owns a real pedal, it will be easy to generate a model from and share around to generate new jc303 presets.

**Comment by RobinSchmidt** [comment]
Yes, that sounds like a great idea. Good to know, that BYOD is the best implementation. Jatin Chowdhury generally does some cool stuff anyway. I recently dabbled with implementing his tape saturation algorithm.

> Indeed 303 and a overdrive is a must!

Indeed, indeed. I also tend to slap a stereo-delay (ping-pong or something) onto it. ...but that's probably not really something, I would integrate directly into the instrument. That can be done in the DAW. For the distortion, it makes more sense to have it directly available in the instrument such that sound-designers can create presets with the right type and amount of distortion already baked in.

**Comment by dfl** [comment]
I have implemented SH-101 style LFO mod... I'll add it as a PR (I've also enhanced the filter, and made a web UI https://dfl.github.io/db303/web/ 🤘)

**Comment by dfl** [comment]
https://github.com/midilab/jc303/pull/34 merry acid xmas 🎅

**Comment by midilab** [comment]
Hi @dfl,

Thanks for the contribution! i love the db303. is this web version using the DiodeFilter already? Diode or DiodeOctave?

I will try to review the PRs asap.

**Comment by dfl** [comment]
yes the web version uses the diode octave model

**Comment by RobinSchmidt** [comment]
Yes - thanks also from my side and merry Christmas from me, too!

**Comment by midilab** [comment]
the modulation is now implemented at develop branch.

@dfl to keep interface and sound tweeking simple and focus on tb303 clone sound i decide to made a minor refactor to your contribution.

* changed the lfo implementation from destination depth based(shs-101 style) to global depth with a 1x destination select(Image-Line TS404 and Arturia Acid V style).
* added volume as destination (Vibrato style like arturia acid v)
* removed pwm(i think it changes too much of original clone idea here, so i striped it out) and pitch(this is subtle and personal, but i dont see much of usage for a 303 clone varying pitch on lfo, but this is open to discusion, as so maybe new modulation destinations)

The current implentation has 4 parameters that can be accesed via DAW plugin parameters area only, still no UI knobs for this feature.
LFO Waveform: 0=Triangle, 1=Saw Up, 2=Saw Down, 3=Square, 4=Random S&H, 5=Noise/Pink
LFO Rate: 0.1 to 20hz
LFO Depth: 0 to 100%
LFO Delstination: 0=filter cutoff, 1=volume(vibrato style)

@RobinSchmidt , @grawlinson, @darrenmothersele, @KBrus 
I would love to hear feedback from all of you about the changes and probrably other new lfo destinations that could be usefull to agregate nice sonic possibilities without going to further from 303 clone proposal.

If anyone cant compile by it self to test the develop branch just ask me then i can upload a version for tests here.

thank you all guys!

**Comment by dfl** [comment]
@midilab season to your taste, however vibrato is certainly LFO to pitch? (LFO to volume is tremolo). Looks like Acid V has a vibrato trigger as part of the step sequencer.

**Comment by midilab** [comment]
>  however vibrato is certainly LFO to pitch?

@dfl you're rigth! my bad... 

> Looks like Acid V has a vibrato trigger as part of the step sequencer.

Didnt know that, maybe it can be usefull, i will add it back pitch as a new destination.

LFO Waveform: 0=Triangle, 1=Saw Up, 2=Saw Down, 3=Square, 4=Random S&H, 5=Noise/Pink
LFO Rate: 0.1 to 20hz
LFO Depth: 0 to 100%
LFO Delstination: 0=filter cutoff, 1=volume(tremolo style), 2=pitch

**Comment by 303inmyheart** [comment]
Greetings, is the Sequencer UI of your screenshot the final version? My idea is that there are 2 types of Sequencers, the OG one (like in the Picture), and a more modern and fluid one Like Acid V, Transitior Bass or the d16 one (still can't pronounce it after all this time:D) In my case (and much more people I know), I/we can't get warm with the og Sequencer.

**Comment by midilab** [comment]
hi @303inmyheart 

The roadmap design shown on the Midilab website represents the final version of part of the sequencer. The idea is to create an exact clone where the sequencer operates in the same way. 

But I feel your pain—programming a 303 isn't much fun; it's a real challenge. So, there's a module (not designed yet) that will allow automatic generation of sequences using certain parameters to create patterns with the push of a button. 

The sequencer and its features will be derived from [Aciduino](https://github.com/midilab/aciduino/blob/master/v2/docs/aciduinoV2-midi_sequencer-user_manual.md#generator-page), which already has these implemented, and they will be ported to JC303.
The generator module (not designed yet):
Acid lines generation is based on:

Probability knobs (0-100%) for: accent, tie, slide
Octave start and octave size to generate
Harmonizer
Length
Shift
Transpose
Fill: amount of notes of total sequence from 0 to 100% for total ramdom note places or use euclidian generator (Bjorklund algorithm)

An Arturia-like step click edit is another module to come after the default clone and auto generator.

**Comment by midilab** [comment]
LFO modulation implemented at #34 and refactored for single destination mode by me at [develop branch](https://github.com/midilab/jc303/tree/develop).
